java-frontend/src/test/files/se/InvariantReturnCheck.java

```diff
@@ -0,0 +1,143 @@
+class A {
+  private int foo(boolean a) { // Noncompliant [[flows=issue1]] {{Refactor this method to not always return the same value.}}
+    int b = 12;
+    if(a) {
+      return b; // flow@issue1
+    } else if(polop()) {
+      return b;  // flow@issue1
+    }
+    return b; // flow@issue1
+  }
+
+  private int foo2(boolean a) {
+    int b = 12;
+    if(a) {
+      return b;
+    }
+    return b - 1;
+  }
+  int foo3(boolean a) {
+    int b = 12;
+    if(a) {
+      return b;
+    }
+    return b;  // false negative : caching of program states because of liveness of params makes the second path unexplored.
+  }
+
+  private String foo4(boolean a) { // Noncompliant
+    String b = "foo";
+    if(a) {
+      return b;
+    }
+    return b;
+  }
+
+  void plop() {
+    a;
+  }
+
+  private int doSomething() {
+    System.out.println("");
+    return 42;
+  }
+  private int getConstant2() {
+    return 42;
+  }
+
+  private A() {}
+
+  String constructComponentName() {
+    synchronized (getClass()) {
+      return base + nameCounter++;
+    }
+  }
+
+  java.util.List<String> f() {
+    System.out.println("");
+    if(bool) {
+      System.out.println("");
+    }
+    return new ArrayList<String>();
+  }
+  java.util.Map<String> f() {
+    java.util.Map<String> foo = new java.util.HashMap<String>();
+    if(bool) {
+      return foo;
+    }
+    return foo;
+  }
+
+  private boolean fun(boolean a, boolean b) { // Noncompliant
+    if(a) {
+      return a;
+    }
+    if(b) {
+      return b;
+    }
+    return true;
+  }
+  private boolean fun2(boolean a, boolean b) {
```

Comment:
for documentation purpose, I would add the following method that you explained already:

```
private boolean fun3(boolean a, boolean b) { // Compliant FN
  if(a) {
    return !b;
  }
  return !b;
}
```